src/sbsign/sbsign.c

```diff
@@ -347,33 +243,256 @@ static int verb_sign(int argc, char *argv[], void *userdata) {
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to set PKCS7 content: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        if (PKCS7_add_certificate(p7, certificate) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to set PKCS7 certificate: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        int x509_mdnid = 0;
+        if (X509_get_signature_info(certificate, &x509_mdnid, NULL, NULL, NULL) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to get X509 digest NID: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        const EVP_MD *md = EVP_get_digestbynid(x509_mdnid);
+        if (!md)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to get digest algorithm via digest NID");
+
+        _cleanup_(PKCS7_SIGNER_INFO_freep) PKCS7_SIGNER_INFO *si = PKCS7_SIGNER_INFO_new();
+        if (!si)
+                return log_oom();
+
+        if (PKCS7_SIGNER_INFO_set(si, certificate, private_key, EVP_get_digestbynid(x509_mdnid)) <= 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to configure signer info: %s",
+                                        ERR_error_string(ERR_get_error(), NULL));
+
+        if (PKCS7_add_attrib_smimecap(si, sk_X509_ALGOR_new_null()) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to add smimecap signed attribute to signer info: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        if (PKCS7_add_attrib_content_type(si, NULL) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to add content type signed attribute to signer info: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        _cleanup_(ASN1_TIME_freep) ASN1_TIME *time = NULL;
+        r = asn1_timestamp(&time);
+        if (r < 0)
+                return r;
+
+        if (PKCS7_add0_attrib_signing_time(si, time) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to add signing time signed attribute to signer info: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        ASN1_OBJECT *idc = OBJ_txt2obj(SPC_INDIRECT_DATA_OBJID, /* no_name= */ true);
+        if (!idc)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to get SpcIndirectDataContent object: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        if (PKCS7_add_signed_attribute(si, NID_pkcs9_contentType, V_ASN1_OBJECT, idc) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to add signed attribute to pkcs7 signer info: %s",
+                                        ERR_error_string(ERR_get_error(), NULL));
+
+        if (PKCS7_add_signer(p7, si) == 0)
```

Comment:
Needs to be `TAKE_PTR(si)` as the destructor of p7 will take ownership

Reply:
Only on success, so that's why we do `TAKE_PTR()` right after this